neuralpp/util/z3_util.py

```diff
@@ -0,0 +1,17 @@
+from z3 import Solver
+from typing import Any
+from copy import copy
+
+
+def z3_merge_solvers(solver1: Solver, solver2: Solver) -> Solver:
+    """  Make a new solver with constraints from both solver1 and solver2. This function does not modify arguments. """
+    result = copy(solver1)
+    result.append(solver2.assertions())
+    return result
+
+
+def z3_add_solver_and_literal(solver: Solver, constraint: Any) -> Solver:
+    """  Make a new solver from the older solver and the added constraint. This function does not modify arguments. """
+    result = copy(solver)
+    result.add(constraint)
```

Comment:
Fun fact: looks like `solver.append`, `solver.insert`, and `solver.add` are alias of the [same method: `solver. assert_exprs`](https://github.com/Z3Prover/z3/blob/master/src/api/python/z3/z3.py#L7014-L7049)